backend/app/services/cluster_service.py

```
@@ -0,0 +1,108 @@
+            "created_at": datetime.now(timezone.utc),
+        })
+
+    # 5. Atomic Batch Creation
+    created_bookings = booking_repo.create_batch_atomic(
+        booking_ids=booking_ids,
+        capacity=capacity,
+        data_list=data_list
+    )
+
+    if created_bookings is None:
+        raise ConflictError("Insufficient capacity for the entire village cluster")
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)